repose-aggregator/components/filters/openstack-identity-v3/src/test/scala/org/openrepose/filters/openstackidentityv3/OpenStackIdentityV3HandlerTest.scala

```diff
@@ -159,6 +159,23 @@ class OpenStackIdentityV3HandlerTest extends FunSpec with BeforeAndAfter with Ma
       headers.keySet() should not contain HeaderName.wrap("X-Impersonator-Id")
     }
 
+    it("should non-destructively add the x-roles header") {
+      when(identityAPI.validateToken("123456", None)).thenReturn(
+        Try(new AuthenticateResponse("1", "2", List(), None, None, Option(List(ServiceForAuthenticationResponse(List(Endpoint("foo", None, None, None, "http://www.notreallyawebsite.com"))))),
+          Option(List(Role("1", "admin"))), UserForAuthenticateResponse(null, None, None, None, None, Some("ORD")))))
+      val mockRequest = new MockHttpServletRequest()
+      mockRequest.setHeader("X-Subject-Token", "123456")
+      identityConfig.setForwardGroups(false)
+      identityConfig.setValidateProjectIdInUri(null)
+      identityV3Handler = new OpenStackIdentityV3Handler(identityConfig, identityAPI)
+      identityV3Handler.handleRequest(mockRequest, mockServletResponse).requestHeaderManager.headersToAdd should contain(
+        Entry(
+          HeaderName.wrap(OpenStackServiceHeader.ROLES.toString),
+          JavaConversions.setAsJavaSet(Set("admin")))
+      )
+      identityV3Handler.handleRequest(mockRequest, mockServletResponse).requestHeaderManager.headersToRemove should not contain HeaderName.wrap(OpenStackServiceHeader.ROLES.toString)
```

Comment:
Yes. Good call.

Well, kind of. So since this filter is adhering to the old pattern which leverages the FilterDirector, no changes are actually made to the request itself. Plus I don't want to test the FilterDirector here. So I'll assume that the FilterDirector works. In that case, since the output of doFilter is a FilterDirector, I can only inspect that. So I just have to verify that the FilterDirector is not removing the header, but is still adding it.

Two calls to handleRequest should not have been called. I'll fix that.